### DISCUSSION TITLE
0.97.0.3 for MC 1.14.*/1.15.*/1.16.*/1.17

### DISCUSSION THREAD

**Discussion by LlmDl** [TownyAdvanced/Towny#5002] [Pre-Release Discussions]
- Fix tree/mushroom structure growth on plot borders not being tested correctly.
  - Update to add compatibility for MC 1.17, courtesy of Warriorrrr and LlmDL with PR #5001.
    - Should account for most of the changes so far. More to come as the API is fleshed out by Spigot.
    - Revert on unclaim feature will probably not like servers using the World Height data pack.
    - Config and world files will update their lists of blocks-ignored-from-regeneration automatically.

----

<details><summary>Cumulative changes since 0.97.0.0</summary>
<details><summary>0.97.0.3</summary>

  - Fix SQL resident loading new jail properties hitting NPEs
    - Closes #4990.  
  - Translate es-419 strings (0.107), courtesy of EuGNosii with PR #4983.
  - Update zh-TW.yml, courtesy of shou692199 with PR #4992.
  - Town and Nation lists are now cached, able to be updated after 10 minutes.
  - Add LuckPerms Contexts, courtesy of Warriorrrr with PR #4966.
    - Adds towny contexts to LuckPerms. At the moment this is very similar to our townyperms.yml, but the difference is that these can be used together with other contexts (per world permissions for example).
    - Includes Resident, Mayor, King, InsideTown, InsideOwnTown, InsideOwnPlot.
  - New Config Option: plugin.interfacing.luckperms_contexts
    - Default: true
    - If enabled, Towny contexts will be available in LuckPerms.
  - Map HUD, courtesy of Warriorrrr with PR #4996.
  - New Command: /towny map hud
    - Activates the map in the player's scoreboard.
</details>
<details><summary>0.97.0.2</summary>

  - Validate /plot group fs input, courtesy of Warriorrrr with PR #4968.
    - Closes #4960.
  - Fix addon api exception in the towny command, courtesy of Warriorrrr with PR #4967.
    - Closes #4965.
  - Shade in slf4j, fixing ClassNotFoundException: org.slf4j.LoggerFactory
    - Closes #4950.
  - Update es-419, courtesy of EuGNosii with PR #4964.
  - Fix towns that already have a nation being able to be sent nation invites.
    - Closes #4963.
  - Fix nations not losing their spawn point.
    - Closes #4961.
  - Fix command always being null, courtesy of Warriorrrr with PR #4974.
    - Closes #4971.
  - Add /ta command for getting exact claimedat time, courtesy of Warriorrrr with PR #4911.
  - New Command: /ta plot claimedat
    - Shows when the plot was claimed.
  - Automatically clean up unused NPC residents, courtesy of Warriorrrr with PR #4972.
  - Parse hex colours in resident titles in the status screens.
    - Closes #4977.
  - Jail Overhaul Project (Begun in January 2021):
    - New Jail Object and JailUtil.
      - Jails are stored in the database in the towny\data\jails\ folder.
      - Towns no longer stores a list of jailspawns.
      - Old town jailspawns will be converted into new Jail objects on first load.
        - Jailed residents will not keep their jailed status, getting a clean slate to start from.
      - Jailing and unjailing code is much better than it used to be.
      - The jailing system is very much the same still, but there are many new nice things.
    - Add Jail Book. 
      - The book is given to players when they are jailed.
      - It explains to the jailed player what they can do to get out of jail, based on how jails have been configured on the server.
      - Closes #4686.
    - Jail plots can now have more than one "cell" or spawn location.
      - SpawnPoints particles are used to show the jail cell spawns.
      - Cells are added/removed using /plot jailcell add|remove
      - Closes #3106.
    - The jailed message now includes who ran the jail command.
      - Closes #4096.
    - Jail plots can now have PVP disabled in them (configurable.)
      - Closes #4659.
    - Chorus Fruit are now lumped in with the Ender Pearl in a new config option for whether teleport items can be used by jailed players.
      -  Closes #4903.
    - Add optional logout protection for newly-jailed players.
      - Newly jailed players that logout before they are teleported to jail will be killed, when the feature is activated.
      - The player is warned of this to prevent surprises.
      - Similar to combat-logging plugins.
    - Add new-player jail-immunity.
      - Players who first joined the server cannot be jailed for their first hour of playtime (configurable.)
    - Add Town PrimaryJail.
      - The PrimaryJail is the default jail for a town.
      - It is used when the /t jail NAME command is used (with no jail/cell/hours specified.)
      - It is used when outlaw and enemies are jailed by being killed.
      - PrimaryJail is set using /t set primaryjail.
    - New Permission Nodes:
      - towny.command.plot.jailcell - allows for adding/removing of jail cells in a jail plot.
      - towny.command.town.jail
      - towny.command.town.jail.list
      - towny.command.town.unjail
      - towny.command.town.set.primaryjail
      - towny.command.plot.jailcell
    - Removed Permission Node: - towny.command.town.toggle.jail
    - New Commands: 
      - /plot jailcell add - Adds a jail cell to a jail plot, where the player is stood.
      - /plot jailcell remove - Removes a jail cell from a jail plot, where the player is stood.
      - /town jail list - Shows jail number, name, coord, cellcount and which jail is the primary jail.
      - /town jail {name} - Jails the given player for 1 hour, must be a resident of your own town.
      - /town jail {name} {hours} - Jails the given player for the given hours.
      - /town jail {name} {hours} {jail} - Jails the given player for the given hours, in the given jail plot (which is a number.)
      - /town jail {name} {hours} {jail} {cell} - Jails the given player for the given hours, in the given jail plot and jail cell (which are both numbers.)
      - /town unjail {name} - Unjails someone in your town's jail.
      - /town set primaryjail - Sets your town's primary jail.
    - Removed Command: /town toggle jail
    - New Config Options:
      - jail.outlaw_jail_hours
        - Default: 5
        - How many hours an attacking outlaw will be jailed for.
      - jail.do_jail_plots_deny_pvp
        - Default: false
        - When true, jail plots will prevent any PVP from occuring. Applies to jailed residents only.
      - jail.prevent_newly_jailed_players_logging_out
        - Default: false
        - When true, Towny will prevent a person who has been jailed by their mayor/town from logging out if they do log out they will be killed first, ensuring they respawn in the jail.
      - jail.new_player_immunity
        - Default: 1h
        - How long do new players have to be on the server before they can be jailed?
      - jail.jail_allows_teleport_items
        - Default: false
        - If true jailed players can use items that teleport, ie: Ender Pearls & Chorus Fruit, but are still barred from using other methods of teleporting.
        - Closes #4903
    - Removed Config Option: jail.outlaw_jail_days
  - Bump Language Files to 0.107.
</details>
<details><summary>0.97.0.1</summary>

  - Fix german.yml, courtesy of relkai with PR #4924.
  - API: Add TownyAPI.getResidentOrNull(TownBlock).
    - Gets the resident who owns the townblock or null if unowned.
    - Best used after TownBlock.hasResident();
  - API: Add TownyAPI.getTownOrNull(TownBlock).
    - Gets the town who owns the townblock or null if wilderness.
    - Best used after TownBlock.hasTown();
  - Fix /n set king declaring that the capital has too many residents.
  - Fix potential console spam from adventure, courtesy of Warriorrrr with PR #4939.
  - bStats to 2.2.1 (was 1.x), Shade it; make isSpigot() private, rename it. Courtesy of TheFlagCourier with PR #4934.
  - Replace deprecated methods on TownyLogger, courtesy of EuGNosii with PR #4933.
  - Update language file es-419.yml, courtesy ofBlock commands from being ran by outsiders #4953#4910.
  - Add Command Addon API, courtesy of Warriorrrr with PR #4886.
    - Adds a new command addon API, that allows plugin developers to add their own sub commands to Towny's existing commands.
  - Replace player.sendMessage() with TownyMessaging.sendMessage().
  - Undo PR #4933 to retain backwards compatibility.
  - Fix no feedback when using a /plot command in the wilderness.
  - Fix regression: null message shown from /t spawn and /n spawn when used from the wilderness.
  - API: Add WorldCoord.getTownOrNull().
    - Gets the town at that WorldCoord or null if it is in the wilderness.
    - Best used after WorldCoord.hasTownBlock();
  - Fix long town/nation names being able to break the TownyFormatter/ChatTools.
    - Closes #4942.
  - Fix sheep being protected from shearing in the wilderness.
  - Improve /town claim auto, works again but doesn't always claim everything it could in one use. 
  - Fix tamed dogs not being hittable inside towns, courtesy of Warriorrrr with PR #4955.
    - Closes #4954.
  - Fix a few tab completion issues with the addon API, courtesy of Warriorrrr with PR #4943.
  - Add commands from being ran by outsiders, courtesy of EuGNosii with PR #4953.
    - Closes #4951.
  - New Config Option: global_town_settings.town_command_blacklisting.own_town_and_wilderness_limited_commands
    - Default: sethome,home
    - This allows the usage of blacklisted commands only in the player's town and the wilderness (essentially blocking commands from being ran by tourists/visitors.)
    - Players with the towny.globally_welcome permission node are not going to be limited by this list.
    - Commands have to be on town_command_blacklisting.town_blacklisted_commands, else this is not going to be checked.
  - Include tb_prefix in column drop check, courtesy of Parsonswy with PR #4952 (First-Time Contributor!)
    - For the mysql servers that could never get the old columns dropped, this will fix that, and prevent the massive error logs that resulted from it.
    - Closes #4935.
  - Bump language files to 0.106.
</details>
</details>

----

<table align=center>
<tr><th colspan=4><h3>Towny Sponsors<br><h4>I want to give a big thank you to all of my sponsors:<br>These are the people that help to make Towny's support and development as active as it is,<br>and who supported me during this pre-release of Towny. </th></tr>
<tr><td>MrLogano</td><td>Momshroom</td><td>Cobrex1</td><td> KarlOfDuty </td></tr>
<tr><td>scatterspell</td><td>gerardvanschip</td><td>partydev</td><td> DrprofLuigi </td></tr>
<tr><td>Tobza </td><td>SmallSansSerif </td><td>TheNylox</td><td>diontimmer </td></tr>
<tr><td>trevor1097</td><td>benevolent2012</td><td>bretomart</td><td>HunSolos</td></tr>
<tr><td>0xBit-MC</td><td>specs1396</td><td>treeDerevo</td><td>Scottlander100</td></tr>
<tr><td>teozfrank</td><td>ItsSynchro</td><td>Advanced-Kind-MC</td><td>MidnightSugar</td></tr>
<tr><td>0liverGMC</td><td>Microhazard</td><td>mhiaxo</td><td>TheFlagCourier</td></tr>
<tr><td colspan=4>and 2 private sponsors.</td></tr>
<tr><td colspan=4 align=center><h3><a href=https://github.com/sponsors/LlmDl>If you want to support the developer consider becoming a sponsor.</a><br><h6>(It's just like Patreon but 100% goes to the developer.)</td></tr>
</table>

<table align=center>
<th colspan=3> Important Links </th>
<tr align=center><td><a title="How Towny Works" href="https://github.com/TownyAdvanced/Towny/wiki/How-Towny-Works">How Towny 
Works</a></td><td><a title="Install Guide" href="https://github.com/TownyAdvanced/Towny/wiki/Installation">Towny Install Guide</a></td><td><a title="Updating Towny" href="https://github.com/TownyAdvanced/Towny/wiki/Updating-Towny">Towny Update Guide</a></td>
</tr>
<tr align=center>
<td><a title="Towny Changelog" href="https://raw.githubusercontent.com/TownyAdvanced/Towny/master/resources/ChangeLog.txt" target="_blank" rel="noopener">Complete Changelog</a> </td><td><a title="Default Config Files" href="https://github.com/TownyAdvanced/Towny/wiki/Config-Files">Default Config Files</a></td><td><a title="Commands/Permissions/Placeholders" href="https://github.com/TownyAdvanced/Towny/wiki/Reference">Commands/Permissions<br>Placeholders</a></td>
</tr>
</table>

### ⏬ Download available as a .jar file in the Assets section below:

<hr /><em>This discussion was created from the release <a href='https://github.com/TownyAdvanced/Towny/releases/tag/0.97.0.4'>0.97.0.3 for MC 1.14.*/1.15.*/1.16.*/1.17</a>.</em>

**Comment by BigTallahasee** [comment]
Great work guys keep it up. Is there anyway I could get some assistance in setting up perms for a bungee server

**Reply by LlmDl** [reply]
Everything is already set up in the townyperms.yml. If you have questions head over to the discord.
